File: backend/routers/ipos.py
from typing import Any

from fastapi import APIRouter, HTTPException

from lib.convex import ConvexUnavailable, call_convex
from models.ipo import IPO, IpoGroups, IposResponse


router = APIRouter()


def _map_ipo(document: dict[str, Any]) -> IPO:
    return IPO(
        id=str(document.get("id") or document.get("_id") or document["sourceId"]),
        name=document["name"],
        category=document["category"],
        status=document["status"],
        price=document.get("price"),
        gmp=document.get("gmp"),
        gmpPercentage=document.get("gmpPercentage"),
        estimatedListing=document.get("estimatedListing"),
        estimatedListingPercentage=document.get("estimatedListingPercentage"),
        expectedProfit=document.get("expectedProfit"),
        profitType=document.get("profitType"),
        lotSize=document.get("lotSize"),
        issueSize=document.get("issueSize"),
        subscription=document.get("subscription"),
        biddingStartDate=document.get("biddingStartDate"),
        biddingEndDate=document.get("biddingEndDate"),
        sourceUrl=document.get("sourceUrl"),
        sourceId=document["sourceId"],
        scrapedAt=document["scrapedAt"],
        updatedAt=document["updatedAt"],
        lastSeenAt=document.get("lastSeenAt") or document["updatedAt"],
        active=document.get("active", True),
    )


def _unavailable_response() -> IposResponse:
    return IposResponse(
        success=False,
        available=False,
        data=IpoGroups(),
        message="IPO data is unavailable until the Convex deployment is configured.",
    )


@router.get("/ipos", response_model=IposResponse)
async def list_ipos() -> IposResponse:
    try:
        documents = await call_convex("query", "ipos:listIpos", {})
    except ConvexUnavailable:
        return _unavailable_response()

    groups = IpoGroups()
    for document in documents or []:
        ipo = _map_ipo(document)
        if ipo.active:
            getattr(groups, ipo.category).append(ipo)
    groups.mainboard.sort(key=lambda item: item.updatedAt, reverse=True)
    groups.sme.sort(key=lambda item: item.updatedAt, reverse=True)
    return IposResponse(success=True, available=True, data=groups)


@router.get("/ipos/{ipo_id}", response_model=IPO)
async def get_ipo(ipo_id: str) -> IPO:
    try:
        document = await call_convex("query", "ipos:getIpo", {"id": ipo_id})
    except ConvexUnavailable as exc:
        raise HTTPException(status_code=503, detail="IPO data is temporarily unavailable") from exc
    if not document:
        raise HTTPException(status_code=404, detail="IPO not found")
    return _map_ipo(document)

File: backend/routes/ipos.ts
import { Router } from "express";

import { callConvex, ConvexUnavailable } from "../services/convex.js";
import {
  normalizeIpo,
  type ConvexIPODocument,
  type IpoGroups,
  type IposResponse,
} from "../types/ipo.js";

export const iposRouter = Router();

function unavailableResponse(): IposResponse {
  return {
    success: false,
    available: false,
    data: { mainboard: [], sme: [] },
    message: "IPO data is unavailable until the Convex deployment is configured.",
  };
}

iposRouter.get("/", async (_request, response) => {
  try {
    const documents = await callConvex<ConvexIPODocument[]>("query", "ipos:listIpos", {});
    const groups: IpoGroups = { mainboard: [], sme: [] };
    for (const document of documents) {
      const ipo = normalizeIpo(document);
      if (ipo.active) groups[ipo.category].push(ipo);
    }
    groups.mainboard.sort((left, right) => right.updatedAt.localeCompare(left.updatedAt));
    groups.sme.sort((left, right) => right.updatedAt.localeCompare(left.updatedAt));
    const payload: IposResponse = { success: true, available: true, data: groups };
    response.json(payload);
  } catch (error) {
    if (error instanceof ConvexUnavailable) {
      response.json(unavailableResponse());
      return;
    }
    response.status(500).json({ detail: "Unable to load IPO data" });
  }
});

iposRouter.get("/:id", async (request, response) => {
  try {
    const document = await callConvex<ConvexIPODocument | null>("query", "ipos:getIpo", {
      id: request.params.id,
    });
    if (!document) {
      response.status(404).json({ detail: "IPO not found" });
      return;
    }
    response.json(normalizeIpo(document));
  } catch (error) {
    if (error instanceof ConvexUnavailable) {
      response.status(503).json({ detail: "IPO data is temporarily unavailable" });
      return;
    }
    response.status(500).json({ detail: "Unable to load IPO data" });
  }
});